### PR TITLE
Replaced all references to DateTime.Now and DateTime.UtcNow with DateTimeProvider

### PR DESCRIPTION
**Pull request by maciejzaleski**
This change is a prerequisite for additional configuration option which will allow node to use current time which will be different to the current system time. This functionality will be used in testing.

### CODE REVIEW

**Comment by bokobza** [comment]
I can see these in the build
```
[C:\projects\fullnode\src\Stratis.SmartContracts.Core.Validation.Tests\Stratis.SmartContracts.Core.Validation.Tests.csproj]
TestFramework\BddSpecification.cs(30,36): error CS0103: The name 'DateTimeProvider' does not exist in the current context [C:\projects\fullnode\src\Stratis.Bitcoin.Tests.Common\Stratis.Bitcoin.Tests.Common.csproj]
TestFramework\BddSpecification.cs(39,38): error CS0103: The name 'DateTimeProvider' does not exist in the current context [C:\projects\fullnode\src\Stratis.Bitcoin.Tests.Common\Stratis.Bitcoin.Tests.Common.csproj]
TestFramework\BddSpecification.cs(40,40): error CS0103: The name 'DateTimeProvider' does not exist in the current context [C:\projects\fullnode\src\Stratis.Bitcoin.Tests.Common\Stratis.Bitcoin.Tests.Common.csproj]
TestFramework\BddSpecification.cs(118,40): error CS0103: The name 'DateTimeProvider' does not exist in the current context [C:\projects\fullnode\src\Stratis.Bitcoin.Tests.Common\Stratis.Bitcoin.Tests.Common.csproj]
```

**Comment by bokobza** [comment]
I'll start the review.

This test is failing though:
image

File: src/Stratis.Bitcoin.Features.Wallet.Tests/WalletManagerTest.cs [hunk 1]
```diff
@@ -377,7 +377,7 @@ public void RecoverWalletWithEqualInputAsExistingWalletRecoversWallet()
                 dataFolder, new Mock<IWalletFeePolicy>().Object, new Mock<IAsyncLoopFactory>().Object, new NodeLifetime(), DateTimeProvider.Default, new ScriptAddressReader());
 
             // Try to recover it.
-            Wallet recoveredWallet = walletManager.RecoverWallet(password, walletName, deletedWallet.mnemonic.ToString(), DateTime.Now.AddDays(1), passphrase);
+            Wallet recoveredWallet = walletManager.RecoverWallet(password, walletName, deletedWallet.mnemonic.ToString(), DateTimeProvider.Default.GetUtcNow().AddDays(1), passphrase);
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet.Tests/WalletManagerTest.cs]: Ideally you should use the same IDateTimeProvider:
```
IDateTimeProvider datetimeProvider = DateTimeProvider.Default;
var walletManager = new WalletManager(this.LoggerFactory.Object, KnownNetworks.StratisMain, chain, NodeSettings.Default(), new Mock<WalletSettings>().Object,
                dataFolder, new Mock<IWalletFeePolicy>().Object, new Mock<IAsyncLoopFactory>().Object, new NodeLifetime(), datetimeProvider, new ScriptAddressReader());

// Try to recover it.
Wallet recoveredWallet = walletManager.RecoverWallet(password, walletName, deletedWallet.mnemonic.ToString(), datetimeProvider.GetUtcNow().AddDays(1), passphrase);
```

same in the other places.

File: src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs [hunk 2]
```diff
@@ -12,7 +12,7 @@ public void GetUtcNowReturnsCurrentUtcDateTime()
         {
             DateTime result = DateTimeProvider.Default.GetUtcNow();
 
-            Assert.Equal(DateTime.UtcNow.ToString("yyyyMMddhhmmss"), result.ToString("yyyyMMddhhmmss"));
+            Assert.Equal(DateTimeProvider.Default.GetUtcNow().ToString("yyyyMMddhhmmss"), result.ToString("yyyyMMddhhmmss"));
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs]: I'm not sure these changes are correct, you're testing the same values here.

File: src/Stratis.Bitcoin.Tests/Utilities/StopwatchTest.cs [hunk 3]
```diff
@@ -23,7 +24,7 @@ public class StopwatchTest
         /// It is expected that all 4 methods will produce roughly the same results.
         /// </para>
         /// <para>
-        /// The first method we use is using two <see cref="System.DateTime.UtcNow"/> calls, one done before and one done after.
+        /// The first method we use is using two <see cref="System.DateTimeProvider.Default.GetUtcNow()"/> calls, one done before and one done after.
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Tests/Utilities/StopwatchTest.cs]: Replace gone bad

File: src/Stratis.Bitcoin.Tests/Utilities/StopwatchTest.cs [hunk 4]
```diff
@@ -44,7 +45,7 @@ public void StopwatchDisposable_MeasuresPerformanceCorrectly()
             int expectedElapsedMs = 0;
             long elapsedTicksByDispStopwatch = 0;
 
-            DateTime startTime = DateTime.UtcNow;
+            DateTime startTime = DateTimeProvider.Default.GetUtcNow();
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Tests/Utilities/StopwatchTest.cs]: Can you define `IDateTimeProvider dateTimeProvider = DateTimeProvider.Default` and use that in the rest of the method?

File: src/Stratis.Bitcoin.Features.Wallet.Tests/WalletManagerTest.cs (hunk 1 above)
Comment (maciejzaleski) [reply src/Stratis.Bitcoin.Features.Wallet.Tests/WalletManagerTest.cs]: I feel that this is a lot of changes for no benefit.

File: src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs (hunk 2 above)
Comment (maciejzaleski) [reply src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs]: Good spot, fixed

File: src/Stratis.Bitcoin.Tests/Utilities/StopwatchTest.cs (hunk 4 above)
Comment (maciejzaleski) [reply src/Stratis.Bitcoin.Tests/Utilities/StopwatchTest.cs]: Changed to use dateTimeProvider

File: src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs [hunk 5]
```diff
@@ -10,9 +10,11 @@ public class DateTimeProviderTest
         [Fact]
         public void GetUtcNowReturnsCurrentUtcDateTime()
         {
+            int timeOffset = -96;
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs]: I think you mixed up your PRs here.

Comment (maciejzaleski) [reply src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs]: Yes, I will quickly move the change to the other PR

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs [hunk 6]
```diff
@@ -5,13 +5,15 @@
...
 using Microsoft.Extensions.Logging.Console;
+using NBitcoin;
 using NLog;
 using NLog.Config;
 using NLog.Extensions.Logging;
 using NLog.Targets;
 using NLog.Targets.Wrappers;
 using Stratis.Bitcoin.Configuration.Settings;
 using Stratis.Bitcoin.Utilities;
+using Target = NLog.Targets.Target;
```
Comment (bokobza) [inline src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs]: Is this necessary?

Comment (bokobza) [reply src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs]: ah because you have NBitcoin's Target, got it.

**Comment by bokobza** [comment]
@maciejzaleski are all the unit and integration tests (but 1) passing?

**Comment by maciejzaleski** [comment]
I have fixed the failing test so all test should be passing now

File: src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs [hunk 7]
```diff
@@ -3,7 +3,7 @@
 using Stratis.Bitcoin.Utilities.Extensions;
 using Xunit;
 
-namespace Stratis.Bitcoin.Tests.Utilities
+namespace NBitcoin
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Tests/Utilities/DateTimeProviderTest.cs]: Please fix the namespace here

**Pull request merged**